In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [9]:
df = pd.read_csv('e_commerce_data.csv', encoding='cp1252')

In [10]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [21]:
df.shape

(541909, 8)

In [17]:
df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

In [29]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [31]:
df[df.CustomerID.isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

In [33]:
df.drop('CustomerID', axis = 1, inplace = True)

In [40]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice         Country  
0  12/1/2010 8:26       2.55  United Kingdom  
1  12/1/2010 8:26       3.39  United Kingdom  
2  12/1/2010 8:26       2.75  United Kingdom  
3  12/1/2010 8:26       3.39  United Kingdom  
4  12/1/2010 8:26       3.39  United Kingdom

In [35]:
df[df.Description.isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

               Country  
622     United Kingdom  
1970    United Kingdom  
1971    United Kingdom  
1972    United Kingdom  
1987    United Kingdom  
...                ...  
535322  United Kingdom  
535326  United Kingdom  
535332  United Kingdom  
536981  United Kingdom  
538554  United Kingdom  

[1454 rows x 7 columns]

In [38]:
df[df.StockCode == '22139']

InvoiceNo StockCode                       Description  Quantity  \
106       536381     22139  RETROSPOT TEA SET CERAMIC 11 PC         23   
622       536414     22139                               NaN        56   
6392      536942     22139                            amazon        15   
6885      536982     22139  RETROSPOT TEA SET CERAMIC 11 PC         10   
7203      537011     22139                               NaN        -5   
...          ...       ...                               ...       ...   
538411    581405     22139  RETROSPOT TEA SET CERAMIC 11 PC          1   
539531    581439     22139  RETROSPOT TEA SET CERAMIC 11 PC          1   
540441    581486     22139  RETROSPOT TEA SET CERAMIC 11 PC          6   
541387    581498     22139  RETROSPOT TEA SET CERAMIC 11 PC          2   
541726    581570     22139  RETROSPOT TEA SET CERAMIC 11 PC          3   

            InvoiceDate  UnitPrice         Country  
106      12/1/2010 9:41       4.25  United Kingdom  
622     12/1/2010 11:52       0.00  United Kingdom  
6392    12/3/2010 12:08       0.00  United Kingdom  
6885    12/3/2010 14:27      11.02  United Kingdom  
7203    12/3/2010 15:38       0.00  United Kingdom  
...                 ...        ...             ...  
538411  12/8/2011 13:50       4.95  United Kingdom  
539531  12/8/2011 16:30      10.79  United Kingdom  
540441   12/9/2011 9:38       4.95  United Kingdom  
541387  12/9/2011 10:26      10.79  United Kingdom  
541726  12/9/2011 11:59       4.95         Germany  

[993 rows x 7 columns]

In [41]:
mapping = df.dropna(subset = ['Description']).drop_duplicates('StockCode').set_index("StockCode")['Description']

In [44]:
mapping

StockCode
85123A     WHITE HANGING HEART T-LIGHT HOLDER
71053                     WHITE METAL LANTERN
84406B         CREAM CUPID HEARTS COAT HANGER
84029G    KNITTED UNION FLAG HOT WATER BOTTLE
84029E         RED WOOLLY HOTTIE WHITE HEART.
                         ...                 
85179a                GREEN BITTY LIGHT CHAIN
23617     SET 10 CARDS SWIRLY XMAS TREE 17104
90214U              LETTER "U" BLING KEY RING
47591b               SCOTTIES CHILDRENS APRON
23843             PAPER CRAFT , LITTLE BIRDIE
Name: Description, Length: 3958, dtype: object

In [56]:
df.Description = df.Description.fillna(df.StockCode.map(mapping))

In [57]:
df.isnull().sum()

InvoiceNo        0
StockCode        0
Description    112
Quantity         0
InvoiceDate      0
UnitPrice        0
Country          0
dtype: int64

In [62]:
df[df.Description.isnull()]

InvoiceNo StockCode Description  Quantity       InvoiceDate  UnitPrice  \
1970      536545     21134         NaN         1   12/1/2010 14:32        0.0   
1987      536549    85226A         NaN         1   12/1/2010 14:34        0.0   
1988      536550     85044         NaN         1   12/1/2010 14:34        0.0   
2024      536552     20950         NaN         1   12/1/2010 14:34        0.0   
2026      536554     84670         NaN        23   12/1/2010 14:35        0.0   
...          ...       ...         ...       ...               ...        ...   
280754    561498     21610         NaN       -14   7/27/2011 14:10        0.0   
281615    561555    37477B         NaN       -11   7/28/2011 10:21        0.0   
281616    561557    37477C         NaN       -31   7/28/2011 10:21        0.0   
346849    567207    35592T         NaN         4   9/19/2011 11:01        0.0   
497301    578360    84971L         NaN         2  11/24/2011 10:36        0.0   

               Country  
1970    United Kingdom  
1987    United Kingdom  
1988    United Kingdom  
2024    United Kingdom  
2026    United Kingdom  
...                ...  
280754  United Kingdom  
281615  United Kingdom  
281616  United Kingdom  
346849  United Kingdom  
497301  United Kingdom  

[112 rows x 7 columns]

The price is zero for all the 112 rows

In [67]:
invoice_nan = df[df.Description.isnull()]['InvoiceNo']

In [71]:
invoice_nan

1970      536545
1987      536549
1988      536550
2024      536552
2026      536554
           ...  
280754    561498
281615    561555
281616    561557
346849    567207
497301    578360
Name: InvoiceNo, Length: 112, dtype: object

In [73]:
df[df.InvoiceNo.isin(invoice_nan)]

InvoiceNo StockCode Description  Quantity       InvoiceDate  UnitPrice  \
1970      536545     21134         NaN         1   12/1/2010 14:32        0.0   
1987      536549    85226A         NaN         1   12/1/2010 14:34        0.0   
1988      536550     85044         NaN         1   12/1/2010 14:34        0.0   
2024      536552     20950         NaN         1   12/1/2010 14:34        0.0   
2026      536554     84670         NaN        23   12/1/2010 14:35        0.0   
...          ...       ...         ...       ...               ...        ...   
280754    561498     21610         NaN       -14   7/27/2011 14:10        0.0   
281615    561555    37477B         NaN       -11   7/28/2011 10:21        0.0   
281616    561557    37477C         NaN       -31   7/28/2011 10:21        0.0   
346849    567207    35592T         NaN         4   9/19/2011 11:01        0.0   
497301    578360    84971L         NaN         2  11/24/2011 10:36        0.0   

               Country  
1970    United Kingdom  
1987    United Kingdom  
1988    United Kingdom  
2024    United Kingdom  
2026    United Kingdom  
...                ...  
280754  United Kingdom  
281615  United Kingdom  
281616  United Kingdom  
346849  United Kingdom  
497301  United Kingdom  

[112 rows x 7 columns]

In [76]:
# it has exactly same rows as the missing description values. Also the Unitprice is zero. So, dropping the rows that do not contribute to analysis

In [77]:
df.dropna(inplace = True)

In [79]:
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
Country        0
dtype: int64